src/Integration/CommandInterceptor.cs

```diff
@@ -0,0 +1,77 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Design;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.OLE.Interop;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Infrastructure.VS;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    internal class CommandInterceptor : IOleCommandTarget
+    {
+        private readonly CommandID commandID;
+        private readonly Func<CommandProgression> function;
+        private readonly IThreadHandling threadHandling;
+
+        public CommandInterceptor(CommandID commandID, Func<CommandProgression> function) : this(commandID, function, ThreadHandling.Instance) { }
+
+        internal /* for testing */ CommandInterceptor(CommandID commandID, Func<CommandProgression> function, IThreadHandling threadHandling)
+        {
+            this.commandID = commandID;
+            this.function = function;
+            this.threadHandling = threadHandling;
+        }
+
+        public int QueryStatus(ref Guid pguidCmdGroup, uint cCmds, OLECMD[] prgCmds, IntPtr pCmdText)
+        {
+            return (int)Microsoft.VisualStudio.OLE.Interop.Constants.OLECMDERR_E_NOTSUPPORTED;
+        }
+
+        public int Exec(ref Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
+        {
```

Comment:
I've noticed that sometimes when I replaced ThreadHelper.ThrowIfNotUOnUIThread i used RunOnUIThread instead, the only reason I didn't do it here is because it was simpler this way. LMK if thats a problem.

Reply:
It depends where we want to put the responsibility for managing threading.

If we throw here, we're effectively saying it is the responsibility of the caller to make sure we are on the correct thread.

If we handle the transition in this class, we're removing the responsibility from the caller.

Here, the only caller will be VS. I would expect that we are always called on the UI thread, so throwing in this case is ok. If it turns out that VS can call the method on a background thread we'll find out pretty soon ;-)